src/Datadog.Trace.DuckTyping/DuckType.Methods.cs

```diff
@@ -0,0 +1,605 @@
+                            {
+                                ILHelpers.WriteLoadArgument(idx, il, false);
+                            }
+
+                            // If the target parameter type is public or if it's by ref we have to actually use the original target type.
+                            targetParamType = targetParamType.IsPublic || targetParamType.IsNestedPublic ? targetParamType : typeof(object);
+                            ILHelpers.TypeConversion(il, proxyParamType, targetParamType);
+
+                            targetMethodParametersTypes[idx] = targetParamType;
+                        }
+                    }
+                }
+
+                // Call the target method
+                if (publicInstance)
+                {
+                    // If the instance is public we can emit directly without any dynamic method
+
+                    // Create generic method call
+                    if (proxyMethodDefinitionGenericArguments.Length > 0)
+                    {
+                        targetMethod = targetMethod.MakeGenericMethod(proxyMethodDefinitionGenericArguments);
+                    }
+
+                    // Method call
+                    if (targetMethod.IsPublic)
+                    {
+                        // We can emit a normal call if we have a public instance with a public target method.
+                        il.EmitCall(targetMethod.IsStatic ? OpCodes.Call : OpCodes.Callvirt, targetMethod, null);
+                    }
+                    else
+                    {
+                        // In case we have a public instance and a non public target method we can use [Calli] with the function pointer
+                        il.Emit(OpCodes.Ldc_I8, (long)targetMethod.MethodHandle.GetFunctionPointer());
+                        il.Emit(OpCodes.Conv_I);
+                        il.EmitCalli(
+                            OpCodes.Calli,
+                            targetMethod.CallingConvention,
+                            targetMethod.ReturnType,
+                            targetMethod.GetParameters().Select(p => p.ParameterType).ToArray(),
+                            null);
+                    }
+                }
+                else
+                {
+                    // A generic method call can't be made from a DynamicMethod
+                    if (proxyMethodDefinitionGenericArguments.Length > 0)
+                    {
+                        throw new DuckTypeProxyMethodsWithGenericParametersNotSupportedInNonPublicInstancesException(proxyMethod);
+                    }
+
+                    // If the instance is not public we need to create a Dynamic method to overpass the visibility checks
+                    // we can't access non public types so we have to cast to object type (in the instance object and the return type).
+
+                    string dynMethodName = $"_callMethod+{targetMethod.DeclaringType.Name}.{targetMethod.Name}";
```

Comment:
Is there any concern about colliding names in the case of method overloads? If we're not keying off these, then maybe not, but I was curious

Reply:
Sounds good, thanks for the response!